urla134d66615_fikr4n_itl_java_tgz-pJ8-example_ExampleJ8/src/org/arabeyes/itl/prayer/PrayerTimeCalc.java

```diff
@@ -0,0 +1,720 @@
+	 * @param date SimpleDate date
+	 * @return imsaak time
+	 */
+	public PrayerTime getImsak(Date date) {
+		
+		CustomMethod tmpConf;
+		DayCouple dc;
+		PrayerTimes temp = new PrayerTimes();
+		
+		tmpConf = new CustomMethod(method);
+		
+		if (method.getFajrInv() != 0) {
+			if (method.getImsaakInv() == 0)
+				tmpConf
+				.setFajrInv((int) (tmpConf.getFajrInv() + Utils.DEF_IMSAAK_INTERVAL));
+			else
+				tmpConf.setFajrInv((int) (tmpConf.getFajrInv() + method
+						.getImsaakInv()));
+			
+		} else if (method.getImsaakInv() != 0) {
+			/* use an inv even if al-Fajr is computed (Indonesia?) */
+			tmpConf.setFajrOffset(tmpConf.getFajrOffset()
+					+ (method.getImsaakInv() * -1));
+			tmpConf.setOffset(true);
+		} else {
+			tmpConf.setFajrAng(tmpConf.getFajrAng() + method.getImsaakAng());
+		}
+		
+		dc = getDayInfo(date, loc.getGmtDiff());
+		getPrayerTimesByDay(tmpConf, dc, temp, InternalTimeType.IMSAK);
+		
+		/* xxxthamer: We probably need to check whether it's possible to compute
+		 * Imsaak normally for some extreme methods first */
+		/* In case of an extreme Fajr time calculation use intervals for Imsaak and
+		 * compute again */
+		if (temp.fajr().isExtreme()) {
+			tmpConf = new CustomMethod(method);
+			if (method.getImsaakInv() == 0) {
+				tmpConf.setFajrOffset(tmpConf.getFajrOffset()
+						- Utils.DEF_IMSAAK_INTERVAL);
+				tmpConf.setOffset(true);
+			} else {
+				tmpConf.setFajrOffset(tmpConf.getFajrOffset()
+						- method.getImsaakInv());
+				tmpConf.setOffset(true);
+			}
+			getPrayerTimesByDay(tmpConf, dc, temp, InternalTimeType.IMSAK);
+		}
+		
+		return temp.fajr();
+	}
+	
+	/* Obtaining the direction of the shortest distance towards Qibla by uMath.sing the
+	 * great circle formula */
+	
+	/**
+	 * Calculate qibla direction.
+	 * @return a Dms object containing qibla direction for the current location
+	 */
+	public Dms getNorthQibla() {
+		return getNorthQibla(loc);
+	}
+
+	/* Obtaining the direction of the shortest distance towards Qibla by uMath.sing the
+	 * great circle formula */
+	
+	/**
+	 * Calculate qibla direction.
+	 * @param loc location where to calculate
+	 * @return a Dms object containg qibla direction 
+	 */
+	public static Dms getNorthQibla(Location loc) {
+		/* xxxthamer: reduce Utils.DEG_TO_RAD usage */
+		double num, denom;
+		num = Math.sin(Utils.degToRad(loc.getDegreeLong())
+				- Utils.degToRad(Utils.KAABA_LONG));
+		denom = (Math.cos(Utils.degToRad(loc.getDegreeLat())) * Math
+				.tan(Utils.degToRad(Utils.KAABA_LAT)))
+				- (Math.sin(Utils.degToRad(loc.getDegreeLat())) * ((Math
+						.cos((Utils.degToRad(loc.getDegreeLong()) - Utils
+								.degToRad(Utils.KAABA_LONG))))));
+		return new Dms(Utils.radToDeg(Math.atan2(num, denom)));
+		
+	}
+
+	private void getPrayerTimesByDay(DayCouple dc, PrayerTimes pt, InternalTimeType type) {
+		getPrayerTimesByDay(method, dc, pt, type);
+	}
+
+	private void getPrayerTimesByDay(Method method, DayCouple dc, PrayerTimes pt,
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This method contains 147 lines of code, not including blank lines or lines with only comments, Java punctuation characters, identifiers, or literals. By comparison, 99% of the methods in the CodeGuru reference dataset contain fewer lines of code. Large methods might be difficult to read and have logic that is hard to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 475-486 into a separate method.